### ISSUE TITLE
Harbor v2.4.0 - oci.image.index push fails if oci.image.manifest hashes are not unique

### ISSUE THREAD

**Issue by zhoudayongdennis** [goharbor/harbor#16718]
If you are reporting a problem, please make sure the following information are provided:

**Expected behavior and actual behavior:**
A clear and concise description of what you expected to happen and what's the actual behavior. If applicable, add screenshots to help explain your problem.

The [OCI Image Index Specification|https://github.com/opencontainers/image-spec/blob/main/image-index.md] does not mention limitation that an oci.image.index can contain only oci.image.manifests with different hash.

with further investigation, there was following error from harbor db pods.

When we try to add OCI index, with same manifests and different annotations, we can see this log in harbor_db POD:

               2022-04-16 19:25:50.249 UTC [6865] ERROR:  duplicate key value violates unique constraint "unique_reference"
               2022-04-16 19:25:50.249 UTC [6865] DETAIL:  Key (parent_id, child_id)=(43, 19) already exists.

This is the table definition:

TABLE artifact_reference
               (
                 id          SERIAL PRIMARY KEY NOT NULL,
                 parent_id   int NOT NULL,
                 child_id    int NOT NULL,
                 child_digest varchar(255) NOT NULL ,
                 platform    varchar(255),
                 urls        varchar(1024),
                 annotations jsonb,
                 FOREIGN KEY (parent_id) REFERENCES artifact(id),
                 FOREIGN KEY (child_id) REFERENCES artifact(id),
                 CONSTRAINT  unique_reference UNIQUE (parent_id, child_id)
               );
CONSTRAINT unique_reference UNIQUE (parent_id, child_id) ← this is the constraint what prevents us to upload index

For WA, 

1. Start shell inside the harbor_db pod
2. Run psql -U postgres
3. Type “\c registry;”
4. Run ALTER TABLE artifact_reference DROP CONSTRAINT unique_reference;
5. Run ALTER TABLE artifact_reference ADD CONSTRAINT unique_reference UNIQUE (parent_id, child_id, platform, annotations);

but, need official fix with more consideration for index base on such scenaio

**Steps to reproduce the problem:**
Please provide the steps to reproduce this problem.

 for the image pushed to harbor registry, if oci.image.index contains more than one oci.image.manifest with the same "digest" (hash).

The following hashes appear more than once in our oci.image.index:

4a368f94cc14267de6b4eaab7fe7a83c247cfe2c7b6c05c996fa5bbac27717e8
368edd22c21f62f93c3d9bdbb0f26f8bf13dd984f423b174dc1ec1967e1b8a58

The oci.image.manifests with the same hash differs in org.opencontainers.image.ref.name (path).

See:

...
        }, {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:4a368f94cc14267de6b4eaab7fe7a83c247cfe2c7b6c05c996fa5bbac27717e8",
            "size": 583,
            "annotations": {
                "org.opencontainers.image.ref.name": "ausf-udm-product/nokia/udm/kubectl:v1.22.1-nano-20211208",
                "com.nokia.ncd.orb.type": "cnfimage"
        }, {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:4a368f94cc14267de6b4eaab7fe7a83c247cfe2c7b6c05c996fa5bbac27717e8",
            "size": 583,
            "annotations": {
                "org.opencontainers.image.ref.name": "ausf-udm-product/corepaas-docker-local.bhisoj70.apac.nsn-net.net/um/kubectl:v1.22.1-nano-20211208",
                "com.nokia.ncd.orb.type": "cnfimage"
            }
        }, {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:368edd22c21f62f93c3d9bdbb0f26f8bf13dd984f423b174dc1ec1967e1b8a58",
            "size": 583,
            "annotations": {
                "org.opencontainers.image.ref.name": "ausf-udm-product/corepaas-docker-local.bhisoj70.apac.nsn-net.net/um/kubectl:v1.23.1-nano-20220128",
                "com.nokia.ncd.orb.type": "cnfimage"
            }
        }, {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:368edd22c21f62f93c3d9bdbb0f26f8bf13dd984f423b174dc1ec1967e1b8a58",
            "size": 583,
            "annotations": {
                "org.opencontainers.image.ref.name": "ausf-udm-product/corepaas-docker-local.bhisoj70.apac.nsn-net.net/lms/belk-efkc/tools/kubectl:v1.23.1-nano-20220128",
                "com.nokia.ncd.orb.type": "cnfimage"
            }
        }, {
            "mediaType": "application/vnd.oci.image.manifest.v1+json",
            "digest": "sha256:368edd22c21f62f93c3d9bdbb0f26f8bf13dd984f423b174dc1ec1967e1b8a58",
            "size": 583,
            "annotations": {
                "org.opencontainers.image.ref.name": "ausf-udm-product/corepaas-docker-local.bhisoj70.apac.nsn-net.net/tools/kubectl:v1.23.1-nano-20220128",
                "com.nokia.ncd.orb.type": "cnfimage"
            }
        }, {
...

**Versions:**
Please specify the versions of following systems.

- harbor version: 2.4.0
- docker engine version: 19.x
- docker-compose version: [z.z.z]

**Additional context:**

- **Harbor config files:** You can get them by packaging `harbor.yml` and files in the same directory, including subdirectory.
- **Log files:** You can get them by package the `/var/log/harbor/` .

**Comment by chlins** [comment]
Could you share why you need to add the index which contains multiple same manifest but different annotations? I think if the manifest is the same, that means they are the completely same image.

**Comment by zhoudayongdennis** [comment]
@chlins,

here is the response for your concern, and this was the real harbor user case. Please consider the improvement to support it.

Yes, they are indeed the same images. But they can have different roles in the manifest listing of an image index. One example is having the same binary image for different platforms. Another example is having the same binary image for different purposes, the purpose being other than platform, and the only way to encode it is the use of annotations. In our case, we establish a special repository in Harbor, into which we re-push a great amount of images from other repositories, among which can appear identical images, and the image index created in this special repository serves as a catalog of all the other images, and we need to maintain the origin of all the re-pushed images. The original repository name is added to the annotations in the re-pushed manifests.

**Comment by chlins** [comment]
@qnetter I think it's a user requirement, please task a look when you have time, thanks.

**Comment by zhoudayongdennis** [comment]
@qnetter, @chlins,

could you feel free to share your assessment base on the request? Any roadmap to support it?

**Comment by qnetter** [comment]
I agree that this can possibly be an invalid optimization in some corner cases. I'm not sure how high the priority of addressing it is, as in how critical those corner cases are...

**Comment by danielzhanghl** [comment]
@qnetter 
when there is large number of  images from different user, there will be likely this kind of issue, 
as OCI standard allows this behavior, there is no way to block that kind of images.

**Comment by danielzhanghl** [comment]
hi maintainers, any plan for this requirement? we could try a  PR if need.

**Comment by zhoudayongdennis** [comment]
@chlins, any further response or feature/enhancement improvement roadmap base on the request from customer?

**Comment by zhoudayongdennis** [comment]
@qnetter, @chlins,

could you feel free to share any further update to move it forward?

**Comment by fk-c-3po** [comment]
I have also encountered this issue when trying to upload this image to Harbor v2.13.2: ghcr.io/berriai/litellm-database:main-v1.77.0-nightly
database 2025-09-12 14:04:23.003 UTC [38671] ERROR:  duplicate key value violates unique constraint "unique_reference"
database 2025-09-12 14:04:23.003 UTC [38671] DETAIL:  Key (parent_id, child_id)=(2195, 2193) already exists.
database 2025-09-12 14:04:23.003 UTC [38671] STATEMENT:  INSERT INTO "artifact_reference" ("parent_id", "child_id", "child_digest", "platform", "urls", "annotations") VALUES ($1, $2, $3, $4, $5, $6) RETURNING "id"

Would have been nice if the database could handle this type of images.
